### PR TITLE
Refactor to improve CassandaraService test coverage

### PR DESCRIPTION
**Pull request by schlosna**
## General
**Before this PR**:
The unit test coverage for `CassandraService` was suboptimal, not catching a potential break in https://github.com/palantir/atlasdb/pull/6074 and did not simulate various cases when Cassandra cluster resizes or encounters various partial failure modes. 

Additionally, the `CassandraServiceTest` can be quite slow locally and [in CI](https://output.circle-artifacts.com/output/job/9c2b7488-e0a5-4ce4-ae37-e0c0194cb7e7/artifacts/3/home/circleci/artifacts/junit/atlasdb-cassandra/test/classes/com.palantir.atlasdb.keyvalue.cassandra.pool.CassandraServiceTest.html) as it must wait for non-existent nodes to timeout, where as we can inject and control the node & address mappings for our test simulations and speed these up.

**After this PR**:
Follow up to https://github.com/palantir/atlasdb/pull/6207 to improve the `CassandraService` test coverage and [`CassandraServiceTest` now runs ~7x faster](https://output.circle-artifacts.com/output/job/a308240a-6b62-4b81-b0de-59f6fc415f2f/artifacts/3/home/circleci/artifacts/junit/atlasdb-cassandra/test/classes/com.palantir.atlasdb.keyvalue.cassandra.pool.CassandraServiceTest.html) compared to [develop](https://output.circle-artifacts.com/output/job/9c2b7488-e0a5-4ce4-ae37-e0c0194cb7e7/artifacts/3/home/circleci/artifacts/junit/atlasdb-cassandra/test/classes/com.palantir.atlasdb.keyvalue.cassandra.pool.CassandraServiceTest.html).

==COMMIT_MSG==
Refactor to improve CassandaraService test coverage
==COMMIT_MSG==

**Priority**: P1

**Concerns / possible downsides (what feedback would you like?)**: 
There are a wide variety of Cassandra rolling restart, partial failure, recovery, and other degraded modes that we don't currently unit test here. I was not planning on building out those additional cases right now given other constraints, but wanted to improve state of the world.

**Is documentation needed?**: no

## Compatibility
**Does this PR create any API breaks (e.g. at the Java or HTTP layers) - if so, do we have compatibility?**:

**Does this PR change the persisted format of any data - if so, do we have forward and backward compatibility?**:

**The code in this PR may be part of a blue-green deploy. Can upgrades from previous versions safely coexist? (Consider restarts of blue or green nodes.)**:

**Does this PR rely on statements being true about other products at a deployment - if so, do we have correct product dependencies on these products (or other ways of verifying that these statements are true)?**:

**Does this PR need a schema migration?**

## Testing and Correctness
**What, if any, assumptions are made about the current state of the world? If they change over time, how will we find out?**:

**What was existing testing like? What have you done to improve it?**:

**If this PR contains complex concurrent or asynchronous code, is it correct? The onus is on the PR writer to demonstrate this.**:

**If this PR involves acquiring locks or other shared resources, how do we ensure that these are always released?**:

## Execution
**How would I tell this PR works in production? (Metrics, logs, etc.)**:

**Has the safety of all log arguments been decided correctly?**:

**Will this change significantly affect our spending on metrics or logs?**:

**How would I tell that this PR does not work in production? (monitors, etc.)**:

**If this PR does not work as expected, how do I fix that state? Would rollback be straightforward?**:

**If the above plan is more complex than “recall and rollback”, please tag the support PoC here (if it is the end of the week, tag both the current and next PoC)**:

## Scale
**Would this PR be expected to pose a risk at scale? Think of the shopping product at our largest stack.**:

**Would this PR be expected to perform a large number of database calls, and/or expensive database calls (e.g., row range scans, concurrent CAS)?**:

**Would this PR ever, with time and scale, become the wrong thing to do - and if so, how would we know that we need to do something differently?**:

## Development Process
**Where should we start reviewing?**:

**If this PR is in excess of 500 lines excluding versions lock-files, why does it not make sense to split it?**:

**Please tag any other people who should be aware of this PR**:
@jeremyk-91
@sverma30
@raiju

### CODE REVIEW

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/pool/CassandraService.java [hunk 1]
```diff
@@ -202,6 +177,69 @@ public ImmutableSet<CassandraServer> refreshTokenRangesAndGetServers() {
...
+                ImmutableSet<CassandraServer> servers = ImmutableSet.copyOf(nodeToDatacentersOnThisTokenRange.keySet());
+                nodes.addAll(servers);
+                nodeToDatacentersOnThisTokenRange.forEach((cassandra, datacenter) -> {
+                    String existing = nodeToDatacentersThisRefresh.put(cassandra, datacenter);
+                    if (existing != null && !Objects.equals(datacenter, existing)) {
+                        log.warn(
+                                "Cassandra node mapped to multiple different datacenters",
+                                SafeArg.of("host", cassandra),
+                                SafeArg.of("datacenter1", existing),
+                                SafeArg.of("datacenter2", datacenter));
```
[the remark this answers is not in the document]
Comment (sudiksha27) [reply atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/pool/CassandraService.java:217]: Can / has this actually happened. This seems like something a `tokenRing` should enforce as an invalid case.

File: atlasdb-cassandra/src/test/java/com/palantir/atlasdb/keyvalue/cassandra/pool/CassandraServiceTest.java [hunk 2]
```diff
@@ -241,34 +254,226 @@ public void selectsHostsWithUnknownDatacenterMappingIfAllKnownDatacentersTried()
...
+        });
+    }
+
+    @Test
+    public void testMultiServerGetCassandraServers() throws Exception {
+        Set<CassandraServer> allHosts = ImmutableSet.of(SERVER_1, SERVER_2, SERVER_3);
+        CassandraService cassandra = clientPoolWithServers(allHosts);
+        List<TokenRange> tokenRanges = generateTokenRanges(3, 1, allHosts);
+        assertThat(cassandra.getCassandraServers(tokenRanges)).hasSize(3);
+    }
```
Comment (sudiksha27) [inline atlasdb-cassandra/src/test/java/com/palantir/atlasdb/keyvalue/cassandra/pool/CassandraServiceTest.java:306]: we should add assertions for tokenMap and hostToDatacenterMapping as well

File: atlasdb-cassandra/src/test/java/com/palantir/atlasdb/keyvalue/cassandra/pool/CassandraServiceTest.java [hunk 3]
```diff
@@ -241,34 +254,226 @@ public void selectsHostsWithUnknownDatacenterMappingIfAllKnownDatacentersTried()
...
+    private static List<TokenRange> generateTokenRanges(Set<CassandraServer> nodes, List<Node> endpointNodes) {
+        List<TokenRange> tokenRanges = Multimaps.index(endpointNodes, Node::start).asMap().entrySet().stream()
+                .map(e -> {
+                    Integer start = e.getKey();
+                    String begin = BaseEncoding.base16().encode(Ints.toByteArray(start));
+                    String end = BaseEncoding.base16().encode(Ints.toByteArray(start + 1));
+
+                    List<EndpointDetails> endpointDetails = e.getValue().stream()
+                            .map(n -> endpointDetails(n.server().cassandraHostName(), DC_1, n.rack()))
+                            .collect(Collectors.toList());
```
Comment (sudiksha27) [inline atlasdb-cassandra/src/test/java/com/palantir/atlasdb/keyvalue/cassandra/pool/CassandraServiceTest.java:442]: why not n.dc() ?

File: atlasdb-cassandra/src/test/java/com/palantir/atlasdb/keyvalue/cassandra/pool/CassandraServiceTest.java [hunk 4]
```diff
@@ -241,34 +254,226 @@ public void selectsHostsWithUnknownDatacenterMappingIfAllKnownDatacentersTried()
...
+        assertThat(endpointNodes).hasSize(nodes.size());
+        return generateTokenRanges(nodes, endpointNodes);
+    }
+
+    private static List<TokenRange> generateTokenRanges(Set<CassandraServer> nodes, List<Node> endpointNodes) {
+        List<TokenRange> tokenRanges = Multimaps.index(endpointNodes, Node::start).asMap().entrySet().stream()
+                .map(e -> {
+                    Integer start = e.getKey();
+                    String begin = BaseEncoding.base16().encode(Ints.toByteArray(start));
+                    String end = BaseEncoding.base16().encode(Ints.toByteArray(start + 1));
```
Comment (sudiksha27) [inline atlasdb-cassandra/src/test/java/com/palantir/atlasdb/keyvalue/cassandra/pool/CassandraServiceTest.java:438]: Don't you have to do (start + 1)% partition?

File: atlasdb-cassandra/src/test/java/com/palantir/atlasdb/keyvalue/cassandra/pool/CassandraServiceTest.java [hunk 5]
```diff
@@ -241,34 +254,226 @@ public void selectsHostsWithUnknownDatacenterMappingIfAllKnownDatacentersTried()
...
+        int nodeCount = replicas * partitions + 1;
+        assertThat(nodeCount).isLessThan(255);
+        return IntStream.range(1, nodeCount)
+                .mapToObj(i -> "10." + i + ".0.0")
+                .map(ip -> InetSocketAddress.createUnresolved(ip, DEFAULT_PORT))
+                .map(CassandraServer::of)
+                .collect(Collectors.toSet());
+    }
+
+    private static List<TokenRange> generateTokenRanges(int replicas, int partitions, Set<CassandraServer> nodes) {
```
Comment (sudiksha27) [inline atlasdb-cassandra/src/test/java/com/palantir/atlasdb/keyvalue/cassandra/pool/CassandraServiceTest.java:396]: hmm this is a lot of code and I am afraid this needs tests of its own. I think it might be worth having token ranges hard coded in files and reading from them

File: atlasdb-cassandra/src/test/java/com/palantir/atlasdb/keyvalue/cassandra/pool/CassandraServiceTest.java [hunk 6]
```diff
@@ -241,34 +254,226 @@ public void selectsHostsWithUnknownDatacenterMappingIfAllKnownDatacentersTried()
...
+                    Integer start = e.getKey();
+                    String begin = BaseEncoding.base16().encode(Ints.toByteArray(start));
+                    String end = BaseEncoding.base16().encode(Ints.toByteArray(start + 1));
+
+                    List<EndpointDetails> endpointDetails = e.getValue().stream()
+                            .map(n -> endpointDetails(n.server().cassandraHostName(), DC_1, n.rack()))
+                            .collect(Collectors.toList());
+
+                    List<String> endpoints = nodes.stream()
+                            .map(CassandraServer::cassandraHostName)
```
Comment (sudiksha27) [inline atlasdb-cassandra/src/test/java/com/palantir/atlasdb/keyvalue/cassandra/pool/CassandraServiceTest.java:445]: why is endpoints all the nodes? should it not be e.value()?

File: atlasdb-cassandra/src/test/java/com/palantir/atlasdb/keyvalue/cassandra/pool/CassandraServiceTest.java [hunk 7]
```diff
@@ -241,34 +254,226 @@ public void selectsHostsWithUnknownDatacenterMappingIfAllKnownDatacentersTried()
...
+        assertThat(cassandra.getCassandraServers(tokenRanges.subList(1, 3))).hasSize(2 * replicas);
+        assertThat(cassandra.getCassandraServers(tokenRanges.subList(2, 4))).hasSize(2 * replicas);
+        assertThat(cassandra.getCassandraServers(tokenRanges.subList(3, 5))).hasSize(2 * replicas);
+        assertThat(cassandra.getCassandraServers(tokenRanges.subList(4, 5))).hasSize(replicas);
+
+        // exhaustively check all ranges
+        for (int i = 0; i < partitions; i++) {
+            for (int j = i + 1; j <= partitions; j++) {
+                for (int k = 1; k <= replicas; k++) {
+                    List<TokenRange> ranges = tokenRanges.subList(i, j);
```
Comment (sudiksha27) [inline atlasdb-cassandra/src/test/java/com/palantir/atlasdb/keyvalue/cassandra/pool/CassandraServiceTest.java:330]: sorry why do we have the k loop?

File: atlasdb-cassandra/src/test/java/com/palantir/atlasdb/keyvalue/cassandra/pool/CassandraServiceTest.java [hunk 8]
```diff
@@ -241,34 +254,226 @@ public void selectsHostsWithUnknownDatacenterMappingIfAllKnownDatacentersTried()
...
+        int partitions = 5;
+        Set<CassandraServer> nodes = generateNodes(replicas, partitions);
+        assertThat(nodes).allSatisfy(node -> assertThat(node.reachableProxyIps())
+                .allSatisfy(address -> assertThat(address.isUnresolved()).isTrue()));
+        CassandraService cassandra = clientPoolWithServers(nodes);
+        List<TokenRange> tokenRanges = generateTokenRanges(replicas, partitions, nodes);
+        assertThat(tokenRanges).hasSize(partitions);
+        assertThat(cassandra.refreshTokenRangesAndGetServers(() -> tokenRanges.subList(0, 3)))
+                .hasSize(9)
+                .satisfies(servers -> {
```
Comment (sudiksha27) [inline atlasdb-cassandra/src/test/java/com/palantir/atlasdb/keyvalue/cassandra/pool/CassandraServiceTest.java:375]: quite hard to parse these constants and why they were chosen

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/pool/CassandraService.java [hunk 9]
```diff
@@ -392,6 +430,40 @@ public RangeMap<LightweightOppToken, ImmutableSet<CassandraServer>> getTokenMap(
...
 
+    private void setTokenMap(ImmutableRangeMap<LightweightOppToken, ImmutableSet<CassandraServer>> tokenRing) {
+        // intern to deduplicate equivalent rings to avoid memory overhead of different views
+        ImmutableRangeMap<LightweightOppToken, ImmutableSet<CassandraServer>> newRing =
+                tokenRingInterner.intern(tokenRing);
+        ImmutableRangeMap<LightweightOppToken, ImmutableSet<CassandraServer>> previousRing = tokenMap; // volatile read
+
+        @SuppressWarnings("ReferenceEquality") // interning provides ability to rely on reference equality for changes
+        boolean ringChanged = previousRing != newRing;
+        if (ringChanged) {
```
Comment (sudiksha27) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/pool/CassandraService.java:441]: Do we have any tests for this?

Comment (sudiksha27) [reply atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/pool/CassandraService.java:441]: in fact I think we should break up this PR into 2
1. race condition + tests
2. 2. refactor + tests

**Review by sudiksha27 — COMMENTED** [review]
The change broadly makes sense. I felt the tests were difficult to read and we could also add more tests for the race condition fix.  
Also, given the size of this PR I recommend we break it up into 2.

File: atlasdb-cassandra/src/test/java/com/palantir/atlasdb/keyvalue/cassandra/pool/CassandraServiceTest.java (hunk 7 above)
Comment (schlosna) [reply atlasdb-cassandra/src/test/java/com/palantir/atlasdb/keyvalue/cassandra/pool/CassandraServiceTest.java:330]: k is replication factor to simulate larger cluster, will adjust to make this clearer

File: atlasdb-cassandra/src/test/java/com/palantir/atlasdb/keyvalue/cassandra/pool/CassandraServiceTest.java (hunk 8 above)
Comment (schlosna) [reply atlasdb-cassandra/src/test/java/com/palantir/atlasdb/keyvalue/cassandra/pool/CassandraServiceTest.java:375]: will adjust to make this clearer

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/pool/CassandraService.java (hunk 9 above)
Comment (schlosna) [reply atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/pool/CassandraService.java:441]: will split this up when I have cycles

**Comment by schlosna** [comment]
closing for now
